doc/extensibility/dataflow_example.md

```diff
@@ -0,0 +1,168 @@
+/// Since this is chained, its versions are derived from the source versions
+/// </summary>
+[Export]
+private class MyChainedDataSource : ChainedProjectValueDataSourceBase<string>
+{
+
+    [ImportingConstructor]
+    protected MyChainedDataSource(UnconfiguredProject project)
+        : base(project.Services)
+    {
+    }
+
+    [Import]
+    private MyOriginalDataSource DataSource { get; set; }
+
+    protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<string>> targetBlock)
+    {
+        // The datasource version is derived from the upstream version,
+        // we do not need a version of our own. However, we must produce a value for every supplied value.
+        // Even if our value does not change.
+        this.JoinUpstreamDataSources(this.DataSource); // Must join upstream to avoid deadlocks!
+        var transformBlock = new TransformBlock<IProjectVersionedValue<string>, IProjectVersionedValue<string>>(
+            i => i.Derive(v => v + "New Value!")); // do some processing with this value and produce a new value
+        var firstLink = this.DataSource.SourceBlock.LinkTo(transformBlock, new DataflowLinkOptions { PropagateCompletion = true });
+        transformBlock.LinkTo(targetBlock, new DataflowLinkOptions { PropagateCompletion = true });
+        return firstLink; // disposing first link will cause everything linked to it to dispose
+    }
+}
+```
+
+## Chained Data Source multiple sources
+
+``` Csharp
+/// <summary>
+/// Links to multiple upstream data sources and produces a single value from them.
+/// This is done using <see cref="ProjectDataSources.SyncLinkTo" />.
+/// </summary>
+[Export]
+private class MyChainedDataSource2 : ChainedProjectValueDataSourceBase<string>
+{
+
+    [ImportingConstructor]
+    protected MyChainedDataSource2(UnconfiguredProject project)
+        : base(project.Services)
+    {
+    }
+
+    [Import]
+    private MyOriginalDataSource DataSource1 { get; set; }
+
+    [Import]
+    private MyOtherOriginalDataSource DataSource2 { get; set; }
+
+    protected override IDisposable LinkExternalInput(ITargetBlock<IProjectVersionedValue<string>> targetBlock)
+    {
+        // The datasource version is derived from the upstream version,
+        // we do not need a version of our own. However, we must produce a value for every supplied value.
+        // Even if our value does not change.
+        this.JoinUpstreamDataSources(this.DataSource1, this.DataSource2); // Must join upstream to avoid deadlocks!
+
+        // When joining multiple sources, your transform block takes in a IProjectVersionedValue<Tuple<>>
+        var transformBlock = new TransformBlock<IProjectVersionedValue<Tuple<string, string>>, IProjectVersionedValue<string>>(
+            i => i.Derive(v => v.Item1 + v.Item2 + "New Value!")); // do some processing with this value and produce a new value
+        var firstLink = ProjectDataSources.SyncLinkTo(
```

Comment:
in this example you have only 2 sources to synchronize and there is an SyncLinkTo overload for 2 or 3 parameters. However when you have a IEnumerable of sources with X , an SyncLinkTo  overload that can do it is not straight forward since needs some casting that one would not know right away. Would be nice to have this example too.